lib/mqtt.c

```diff
@@ -143,81 +143,105 @@ static int mqtt_getsock(struct Curl_easy *data,
   return GETSOCK_READSOCK(FIRSTSOCKET);
 }
 
-/* Compute packet length byte */
-static char packetlen_byte(size_t packetlen)
+static int mqtt_encode_len(char *buf, size_t len)
 {
-  DEBUGASSERT(packetlen < 125);
-  return (packetlen - 2) & 0x7f;
+  unsigned char encoded;
+  int i;
+
+  for(i = 0; (len > 0) && (i<4); i++) {
+    encoded = len % 0x80;
+    len /= 0x80;
+    if(len)
+      encoded |= 0x80;
+    buf[i] = encoded;
+  }
+
+  return i;
 }
 
 /* add the passwd to the CONNECT packet */
 static void add_passwd(const char *passwd, const size_t plen,
-                       char *pkt, const size_t start)
+                       char *pkt, const size_t start, int remain_pos)
 {
   /* magic number that need to be set properly */
-  const size_t conn_flags_pos = 9;
+  const size_t conn_flags_pos = remain_pos + 8;
 
   /* set password flag */
   pkt[conn_flags_pos] |= 0x40;
 
   /* length of password provided */
-  pkt[start + 1] = (char)plen;
+  DEBUGASSERT(plen < 65535);
+  pkt[start] = (char)((plen >> 8) & 0xFF);
+  pkt[start + 1] = (char)(plen & 0xFF);
   memcpy(&pkt[start + 2], passwd, plen);
 }
 
 /* add user to the CONN packet */
 static void add_user(const char *username, const size_t ulen,
-                     unsigned char *pkt, const size_t start)
+                     unsigned char *pkt, const size_t start, int remain_pos)
 {
   /* magic number that need to be set properly */
-  const size_t conn_flags_pos = 9;
+  const size_t conn_flags_pos = remain_pos + 8;
 
   /* set username flag */
   pkt[conn_flags_pos] |= 0x80;
   /* length of username provided */
-  pkt[start + 1] = (char)ulen;
+  DEBUGASSERT(ulen < 65535);
+  pkt[start] = (char)((ulen >> 8) & 0xFF);
+  pkt[start + 1] = (char)(ulen & 0xFF);
   memcpy(&pkt[start + 2], username, ulen);
 }
 
-/* Set initial values of CONN packet */
-static void init_connpack(char *packet, size_t packetlen)
+/* add client ID to the CONN packet */
+static void add_client_id(const char *client_id,
+                     char *pkt, const size_t start)
```

Comment:
I think this function should also take the client id length and then you can check its correctness with an assert so that we can be sure we don't slip up in the implementation, as this functions will memcpy() to this buffer with a fixed length.